From 96129035692f21c6826d33c04eeb9927dc6025fc Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Mon, 22 Jan 2024 11:26:46 +0100
Subject: [PATCH 0579/1695] drivers: stm32mp2_pmic: handle over current
 protection

- Setting regulator-over-current-protection now set in HICUP mode
as the default setting from NVMEM is fail-safe mode.
- Allow to override the OCP handler.
- panic on over-current detection in the default handler.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git bb093b50f2f5abc8ec69f4530c505e5ba9bab21b drivers: stm32mp2_pmic: handle over current protection]
Change-Id: I305921e7694bc6ada2981e42a0f6298549f1419b
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/353273
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stpmic2.c | 9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

diff --git a/core/drivers/stpmic2.c b/core/drivers/stpmic2.c
index 913e07df4..0647b11c7 100644
--- a/core/drivers/stpmic2.c
+++ b/core/drivers/stpmic2.c
@@ -470,8 +470,7 @@ TEE_Result stpmic2_regulator_set_prop(struct stpmic2 *pmic, uint8_t id,
 	case STPMIC2_OCP:
 		/* enable mask reset */
 		return stpmic2_register_update(pmic, regul->ocp_reg,
-					       regul->ocp_mask,
-					       regul->ocp_mask);
+					       U(0), regul->ocp_mask);
 	case STPMIC2_PWRCTRL_EN:
 		if (!regul->pwrctrl_cr)
 			return TEE_ERROR_BAD_PARAMETERS;
@@ -630,13 +629,11 @@ static void stpmic2_handle_ocp(struct stpmic2 *pmic, uint8_t reg_index,
 		if (val & BIT(i)) {
 			uint8_t id = (reg_index - 2) * 8 + i;
 
-			EMSG("******************************************");
 			EMSG("Overcurrent detected on %s, disable regulator",
 			     regul_table[id].name);
-			EMSG("******************************************");
 
-			if (stpmic2_regulator_set_state(pmic, id, false))
-				panic();
+			stpmic2_regulator_set_state(pmic, id, false);
+			panic();
 		}
 	}
 }
-- 
2.47.2

